primer-miso/src/Primer/Miso/Util.hs

```diff
@@ -250,3 +287,89 @@ nodeSelectionType =
           THole{} -> True
           TEmptyHole{} -> True
           _ -> False
+
+-- TODO keep an eye out for whether these ever return `Nothing` in practice
+-- the current hypothesis is that, because we already typecheck on initialisation and on performing each action,
+-- the app only ever deals with typechecked expressions
+-- and that it is therefore possible that we could refactor the library to assert this in the Haskell types
+-- note that we try to always work with typechecked expressions in the frontend,
+-- as being able to assume that metadata is always available makes the core application code simpler
```

Comment:
~Can you elaborate a bit on this point in a comment here in the PR thread? This point is probably my only real concern with this PR, and I just want to make sure it's obviously called out in the GitHub UI, so that it's easier to find.~

Never mind, I think it's obvious enough now with my subsequent comments. Let's just not mark this comment as "resolved" so that it remains obvious.

Reply:
My initial thought is that probably we should just change `primer-api` so that it always returns type-checked exprs, ASTs, etc. Obviously that's out of scope for this PR and will require quite a bit of refactoring, as you say.